### PR TITLE
chore: Refactoring API wiring for actions and JS actions to support private entity renaming on EE

### PR DESCRIPTION
**Pull request by ankitakinger**
## Description

Refactoring API wiring for actions and JS actions to support private entity renaming on EE. Also handles a couple of other issues (refer the issue attached below)

#### PR fixes following issue(s)
Fixes [#29762](https://github.com/appsmithorg/appsmith/issues/29762)

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

#### How Has This Been Tested?
- [x] Manual
- [ ] JUnit
- [ ] Jest
- [x] Cypress

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
  - Enhanced action and JS object naming capabilities with new context-aware options.
  - Added support for organizing actions and JS objects within modules.

- **Improvements**
  - Streamlined the process of renaming actions and JS objects to be more intuitive and context-sensitive.

- **Refactor**
  - Updated internal type declarations for consistency and future extensibility.

- **User Interface**
  - Improved UI elements to reflect the new naming and organizational features for actions and JS objects.

- **Bug Fixes**
  - Corrected logic to ensure proper handling of the `isPublic` flag within JS collections.

### CODE REVIEW

File: app/client/src/utils/AppsmithUtils.tsx [hunk 1]
```diff
@@ -198,21 +198,21 @@ export const getDuplicateName = (prefix: string, existingNames: string[]) => {
 export const createNewApiName = (
   actions: ActionDataState,
   entityId: string,
-  key: CreateNewActionKey = CreateNewActionKey.PAGE,
+  key = CreateNewActionKey.PAGE,
 ) => {
   const pageApiNames = actions
-    .filter((a) => a.config[key] === entityId)
+    .filter((a: any) => a.config[key] === entityId)
```
Comment (ApekshaBhosale) [inline app/client/src/utils/AppsmithUtils.tsx:204]: is it possible to add types here @ankitakinger ?

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

File: app/client/src/utils/AppsmithUtils.tsx (hunk 1 above)
Comment (ankitakinger) [reply app/client/src/utils/AppsmithUtils.tsx:204]: This is intentional as of now and a temporary change. I'll be looking into this again.

File: app/client/src/pages/Editor/Explorer/Files/FilesContextProvider.tsx [hunk 2]
```diff
@@ -36,33 +36,31 @@ export const FilesContextProvider = ({
...
-    if (parentEntityType === ACTION_PARENT_ENTITY_TYPE.PAGE) {
-      items.push(
-        ActionEntityContextMenuItemsEnum.SHOW_BINDING,
-        ActionEntityContextMenuItemsEnum.COPY,
-        ActionEntityContextMenuItemsEnum.MOVE,
-      );
-    }
-    return items;
-  }, [parentEntityType]);
+  const defaultMenuItems = [
```
Comment (ashit-rath) [inline app/client/src/pages/Editor/Explorer/Files/FilesContextProvider.tsx:45]: Keep this default outside as the array would be recreated on every re-render thus failing the useMemo and use it like 
```
menuItems = DEFAULT_MENU_ITEMS
```
in line 39 and get rid of `useMemo`

**Review by ApekshaBhosale — APPROVED** [review]

File: app/client/src/utils/AppsmithUtils.tsx [hunk 3]
```diff
@@ -198,21 +198,21 @@
...
 
 export const createNewJSFunctionName = (
   jsActions: JSCollectionData[],
   entityId: string,
-  key: CreateNewActionKey = CreateNewActionKey.PAGE,
+  key = CreateNewActionKey.PAGE,
 ) => {
   const pageJsFunctionNames = jsActions
-    .filter((a) => a.config[key] === entityId)
+    .filter((a: any) => a.config[key] === entityId)
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/utils/AppsmithUtils.tsx]: will be handled separately

File: app/client/src/utils/AppsmithUtils.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/utils/AppsmithUtils.tsx]: will be handled separately

File: app/client/src/pages/Editor/Explorer/Files/FilesContextProvider.tsx [hunk 4]
```diff
@@ -36,40 +44,27 @@ export const FilesContextProvider = ({
...
-  }, [parentEntityType]);
-
   const value = useMemo(() => {
     return {
       canCreateActions,
       editorId,
       parentEntityId,
       parentEntityType,
-      menuItems,
+      menuItems: menuItems || defaultMenuItems,
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Files/FilesContextProvider.tsx:59]: ensured the same

File: app/client/src/api/ActionAPI.tsx [hunk 5]
```diff
@@ -141,11 +142,13 @@ export interface CopyActionRequest {
...
 export interface UpdateActionNameRequest {
-  pageId: string;
+  pageId?: string;
   actionId: string;
-  layoutId: string;
+  layoutId?: string;
   newName: string;
   oldName: string;
+  moduleId?: string;
+  contextType?: ActionContextTypeInterface;
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/api/ActionAPI.tsx:151]: intentional

File: app/client/src/ce/api/JSActionAPI.tsx [hunk 6]
```diff
@@ -35,7 +37,8 @@
   applicationId: string;
   pluginType: PluginType;
   workflowId?: string;
-  contextType?: ActionContextType;
+  contextType?: ActionContextTypeInterface;
+  moduleId?: string;
 }
 
 export interface SetFunctionPropertyPayload {
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/api/JSActionAPI.tsx:44]: intentional

File: app/client/src/entities/Action/index.ts [hunk 7]
```diff
@@ -163,7 +163,7 @@ export interface BaseAction {
   moduleId?: string;
   moduleInstanceId?: string;
   workflowId?: string;
-  contextType?: ActionContextType;
+  contextType?: ActionContextTypeInterface;
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/entities/Action/index.ts:166]: intentional

File: app/client/src/entities/JSCollection/index.ts [hunk 8]
```diff
@@ -24,11 +24,15 @@ export interface JSCollection {
...
-  contextType?: ActionContextType;
+  contextType?: ActionContextTypeInterface;
   // This is used to identify the main js collection of a workflow
   // main js collection is the entrypoint for a workflow
   // cannot be deleted or renamed
   isMainJSCollection?: boolean;
+  displayName?: string;
+  hideEditIconOnEditor?: boolean;
+  hideContextMenu?: boolean;
+  hideContextMenuOnEditor?: boolean;
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/entities/JSCollection/index.ts:35]: intentional

File: app/client/src/ce/api/JSActionAPI.tsx [hunk 9]
```diff
@@ -17,11 +17,13 @@ export interface MoveJSCollectionRequest {
...
 export interface UpdateJSObjectNameRequest {
-  pageId: string;
+  pageId?: string;
   actionCollectionId: string;
-  layoutId: string;
+  layoutId?: string;
   newName: string;
   oldName: string;
+  moduleId?: string;
+  contextType?: ActionContextTypeInterface;
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/api/JSActionAPI.tsx:26]: intentional

File: app/client/src/ee/entities/Engine/actionHelpers.ts [hunk 10]
```diff
@@ -1,6 +1,10 @@
 export * from "ce/entities/Engine/actionHelpers";
+import type { ActionContextType } from "ce/entities/Engine/actionHelpers";
 import { ACTION_PARENT_ENTITY_TYPE as CE_ACTION_PARENT_ENTITY_TYPE } from "ce/entities/Engine/actionHelpers";
 
+export type ActionContextTypeInterface =
+  (typeof ActionContextType)[keyof typeof ActionContextType];
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ee/entities/Engine/actionHelpers.ts:12]: intentional

**Review by ApekshaBhosale — APPROVED** [review]

**Pull request merged**